content/post/2022-05-01-titus-storage-part2.markdown

```diff
@@ -0,0 +1,268 @@
+And you can't just add `--privileged` to [`docker exec`](https://github.com/moby/moby/issues/21984), it must be applied to the container at start time.
+
+`--privileged`? `CAP_SYSADMIN`? `--security-opt apparmor=unconfined`?
+Seems like a lot of security we have to disable to make it work though.
+But this is the opposite of what I want. I want *more* security, not less.
+This is even more complex when we throw in Linux User Namespaces.
+
+## User Namespaces
+
+[Linux user namespaces](http://man7.org/linux/man-pages/man7/user_namespaces.7.html) are a way mapping a UID/GIDs in a container.
+For securing containers, [Titus](https://netflixtechblog.com/evolving-container-security-with-linux-user-namespaces-afbe3308c082) user user namespaces to ensure the `root` in the container is **not** root on the host.
+
+At the time of this writing, user namespaces are *not* enabled by default on [Docker (sometimes calld "rootless docker")](https://docs.docker.com/engine/security/rootless/) or [Podman](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux_atomic_host/7/html-single/managing_containers/index#set_up_for_rootless_containers).
+
+But, if rootless mode is setup in docker, can we *will* run mount?:
+
+```bash-session
+<docker-rootless> $ docker run --cap-add SYS_ADMIN ubuntu mount --verbose -t tmpfs none /mnt
+```
+
+On modern kernels this works!
+Interestingly, we don't even need to disable AppArmor or Seccomp, as they are not enabled by default in rootless mode...
+
+But what if we are OK with running a command as `root` (real root) outside of the container to mount storage *inside* container.
+Why can't we just "run `mount`" for a container, even if it is rootless or has no additional privileges?
+
+## Injecting Mounts in Containers
+
+You *can* run `mount` on behalf of a container!
+What makes the difference between mounting something inside or outside a container?
+How does Linux "know" that something is mounted in a container or not?
+The answer: "The linux mount namespace".
+
+If you can enter the mount namespace first, *then* `mount`, the mount will "be" in the container's mount namespace:
+
+<table>
+<tr><td> Outside the container </td> <td> Inside the container </td></tr>
+<tr><td>
+
+```bash-session
+...
+...
+# P=$(docker inspect -f '{{.State.Pid}}' bdfaa6ac3407)
+# nsenter -t $P -m mount -t tmpfs tmpfs /mnt
+...
+...
+...
+```
+</td><td>
+
+```bash-session
+# docker run -ti ubuntu
+root@bdfaa6ac3407:/# grep mnt /proc/mounts 
+...
+...
+root@bdfaa6ac3407:/# grep mnt /proc/mounts 
+tmpfs /mnt tmpfs rw,relatime,inode64 0 0
+root@bdfaa6ac3407:/# 
+```
+
+</td></tr>
+</table>
+
+While this looks like a good lead, it will *not* work for block devices in `/dev`.
+Why?
+Because as soon as you change into the container's mount namespace, you no longer can see `/dev`.
+
+Even if we could see dev files, or fuse, or network filesystems, the `mount` wouldn't work if we had the **user namespaces** on too.
```

Comment:
This isn't completely true. There are some filesystems that have the flag on to allow for user namespaces (with cap_sys_admin in the user namespace) to mount the filesystem.

Reply:
You are talking these?:
https://man7.org/linux/man-pages/man7/user_namespaces.7.html

```
* /proc (since Linux 3.8)
           * /sys (since Linux 3.8)
           * devpts (since Linux 3.9)
           * [tmpfs(5)](https://man7.org/linux/man-pages/man5/tmpfs.5.html) (since Linux 3.9)
           * ramfs (since Linux 3.9)
           * mqueue (since Linux 3.9)
           * bpf (since Linux 4.4)
           * overlayfs (since Linux 5.11)
```